Name:
**Zachary Gadjiel Breinard G. Que**

Course code and section:  **DATA100 - S17**

Student ID: **12135666**

College: **CCS**

Course: **BS Interactive Entertainment Technology Major in Game Development**

References Used:
https://pandas.pydata.org/docs/index.html

# Zomato Exercise

Zomato is a restaurant aggregator website. The platform provides information, menus and user-reviews of restaurants as well as food delivery options from partner restaurants in select cities. You have been provided 3 datasets:

- zomato.csv contains restaurant information like average_cost_for_two, cuisines
- location.csv contains location of the restaurants like address, city, locality & etc
- ratings.csv contains aggregate rating for each restaurant

The different files are connected using the `id` columns which can be found in the different files. The id column is a unique identifier to represent a restaurant. You can use this to **merge** the different files.

In [1]:
import pandas as pd
import numpy as np

1. Merge the zomato, location & ratings file into a signle dataframe and store it to the `merged_df` variable **(3pts)**

In [2]:
merged_df = None

# read the files 
zomato = pd.read_csv('zomato/zomato.csv')
location = pd.read_csv('zomato/location.csv')
ratings = pd.read_csv('zomato/ratings.csv')

# After checking the csvs, common index is id 

# merge zomato and location 
zomatoAndLocation = pd.DataFrame.merge(zomato, location, on='id')

# merge zomatoAndLocation and ratings
merged_df = pd.DataFrame.merge(zomatoAndLocation, ratings, on='id')

In [3]:
assert (6830, 35)==merged_df.shape

cols = ['id', 'address', 'city', 'city_id', 'country_id', 'latitude',
       'locality', 'locality_verbose', 'longitude', 'zipcode',
       'aggregate_rating', 'rating_color', 'rating_text', 'votes',
       'average_cost_for_two', 'book_url', 'cuisines', 'currency', 'deeplink',
       'events_url', 'featured_image', 'has_online_delivery',
       'has_table_booking', 'include_bogo_offers', 'is_book_form_web_view',
       'is_delivering_now', 'is_table_reservation_supported',
       'is_zomato_book_res', 'menu_url', 'mezzo_provider', 'name',
       'photos_url', 'price_range', 'thumb', 'url']

assert sorted(merged_df.columns) == sorted(cols)

2. Create a function that returns the average aggregate rating for each city. Expected returned value is a Series datatype which contains name of city as index and average aggregate rating as value. **(2pts)**

In [4]:
def exercise_2(merged_df):
    
    index =  merged_df.groupby('city')
    values = index['aggregate_rating'].mean()
    
    return pd.Series(data=values.values, index=values.index, name="average_aggregate_rating")


print(merged_df.groupby('city'))

In [5]:
res_series = exercise_2(merged_df)
assert True == np.isclose(3.302075, res_series[res_series.index == 'Pasay City'],
                          rtol=1e-05, atol=1e-08, equal_nan=False)[0]
assert True == np.isclose(1.890741, res_series[res_series.index == 'Valenzuela City'],
                          rtol=1e-05, atol=1e-08, equal_nan=False)[0]
assert True == np.isclose(1.558333, res_series[res_series.index == 'Malabon City'],
                          rtol=1e-05, atol=1e-08, equal_nan=False)[0]

3. Using the apply function of pandas convert the rating_text to a numeric equivalent. You are to introduce a new column in merged_df called `rating_numeric` which is based on the value from rating_text: **(3pts)**

Excellent = 5 <br>
Very Good = 4 <br>
Good = 3 <br>
Average = 2 <br>
Poor = 1 <br>
Not rated = 0 <br>

In [6]:
def convertRatingTextToNumeric(rating_text):
    if rating_text == "Excellent":
        return 5
    elif rating_text == "Very Good":
        return 4
    elif rating_text == "Good":
        return 3
    elif rating_text == "Average":
        return 2
    elif rating_text == "Poor":
        return 1
    elif rating_text == "Not rated":
        return 0

merged_df['rating_numeric'] = merged_df['rating_text'].apply(convertRatingTextToNumeric)

In [7]:
assert 868 == merged_df[merged_df.rating_numeric == 0].shape[0]
assert True == ('rating_numeric' in merged_df.columns)

4. Create a function which filters the merged dataframe to only contain rows coming in from a specific city. Expected returned value is a dataframe **(2pts)**

In [8]:
def exercise_4(merged_df, city):
    return merged_df[merged_df['city'] == city]

In [9]:

assert 'Makati City' == exercise_4(merged_df, 'Makati City').city.unique()[0]
assert (976, 36) == exercise_4(merged_df, 'Makati City').shape

assert 'Mandaluyong City' == exercise_4(merged_df, 'Mandaluyong City').city.unique()[0]
assert (443, 36) == exercise_4(merged_df, 'Mandaluyong City').shape

assert (0, 36) == exercise_4(merged_df, 'Quezon  City').shape

5. Create a function to filter the dataframe to only contain certain columns passed in the function. Expected returned value is a dataframe **(1pt)**

In [10]:
def exercise_5(merged_df, cols_to_filter):
    return merged_df[cols_to_filter]

In [11]:
cols = ['name', 'cuisines', 'city', 'aggregate_rating']
assert sorted(cols) == sorted(exercise_5(merged_df, cols).columns)
assert (6830, 4) == exercise_5(merged_df, cols).shape

cols = ['average_cost_for_two', 'book_url', 'cuisines', 'currency', 'deeplink',
       'events_url', 'featured_image', 'has_online_delivery',
       'has_table_booking', 'include_bogo_offers', 'is_book_form_web_view']
assert sorted(cols) == sorted(exercise_5(merged_df, cols).columns)
assert (6830, 11) == exercise_5(merged_df, cols).shape


6.  What is the most common restaurant name in the dataset? **(2pts)** Write down your answer and show code used to come up with the answer.

In [12]:
# Get the name with the highest value
mostCommonResto = merged_df['name'].value_counts().idxmax()
print("The most common restaurant is " + mostCommonResto + "\n")

# Checker
commonResto = merged_df['name'].value_counts()
print(commonResto)

The most common restaurant is Starbucks

name
Starbucks                    99
Jollibee                     77
McDonald's                   66
KFC                          54
BonChon Chicken              49
                             ..
Spatzle Euro Market Cafe      1
Corazon                       1
Yelo A Taste Of Snow          1
Islas Pinas                   1
Red Buffalo Wings & Pizza     1
Name: count, Length: 3286, dtype: int64


Answer: The most common restaurant in the dataset is <u><b> Starbucks </b></u>

7. Which restaurant has the most expensive average cost for two? **(2pts)** Write down your answer and show code used to come up with the answer.

In [22]:
# Get row with highest average cost for two
mostExpensiveResto = merged_df.loc[merged_df['average_cost_for_two'].idxmax()]
# print name of that row
print("The most expensive restaurant for two is " + mostExpensiveResto['name'] + "\n")

# Checker
expensiveRestos = merged_df.sort_values(by='average_cost_for_two', ascending=False)
print(expensiveRestos[['name', 'average_cost_for_two']])


The most expensive restaurant for two is The Tasting Room

                                           name  average_cost_for_two
1713                           The Tasting Room                 10000
5      Spiral - Sofitel Philippine Plaza Manila                  6000
1728               Medley Buffet - Okada Manila                  5400
1750  Wolfgang's Steakhouse by Wolfgang Zwiener                  5100
2039  Wolfgang's Steakhouse by Wolfgang Zwiener                  5100
...                                         ...                   ...
6796                           Katherine's Cafe                     0
4714                            The Baking Crew                     0
6482                                   Toritake                     0
6192                                  Sari Salo                     0
6509                                   KapeTayo                     0

[6830 rows x 2 columns]


Answer: The restaurant with the most expensive average cost for two is <u><b> The Tasting Room </b></u>

8. How many restaurants do not have a zipcode? **(1pt)** Write down your answer and show code used to come up with the answer.

In [44]:
# Get all rows with no zipcode
noZipCode = merged_df['zipcode'].isna()
# Print the total number of rows with no zipcode
print("Number of restaurants with no zipcode: " + str(noZipCode.sum()))

# Checker
sortZipCode = merged_df.sort_values(by='zipcode', ascending=False)
withZipCode = merged_df[merged_df['zipcode'].notna()]
countWithZipCode = withZipCode['zipcode'].count()
assert merged_df.shape[0] == countWithZipCode + noZipCode.sum()

Number of restaurants with no zipcode: 6713


Answer: There are <u><b> 6713</b></u> restaurants with no zipcode

9. How many restaurants in the dataset allow online delivery? **(1pt)** Write down your answer and show code used to come up with the answer.

In [51]:
# get has_online_delivery column and count how many are 1s
hasOnlineDelivery = merged_df['has_online_delivery'].sum()
print("Number of restaurants with online delivery: " + str(hasOnlineDelivery) + "\n")

# checker
withOnlineDelivery = merged_df.sort_values(by='has_online_delivery', ascending=False)
print(withOnlineDelivery[['name', 'has_online_delivery']])

Number of restaurants with online delivery: 0

                   name  has_online_delivery
0     Mendokoro Ramenba                    0
4561                FIC                    0
4559          Starbucks                    0
4558         McDonald's                    0
4557      Pancake House                    0
...                 ...                  ...
2274      Kowloon House                    0
2273        Mang Inasal                    0
2272            Janet's                    0
2271           Chowking                    0
6829          Greenwich                    0

[6830 rows x 2 columns]


10. Which locatlity has more expensive food? Poblacion or Salcedo Village? **(3pts)** Write down your answer and show code used to come up with the answer.

In [82]:
# get restos in Poblacion
poblaRestos = merged_df[merged_df['locality'].str.contains("Poblacion")] # check for all localities that contain "Poblacion" i.e. Century City Mall, Poblacion, Makati City 
# get average cost of restos in Poblacion
poblaAvgCost = poblaRestos['average_cost_for_two'].mean()

# get restos in SalcedoVillage
salcedoRestos = merged_df[merged_df['locality'].str.contains("Salcedo Village")] # check for all localities that contain "Salcedo Village" i.e. Ayala Triangle Gardens, Salcedo Village, Makati City
# get average cost of restos in Salcedo Village
salcedoAvgCost = salcedoRestos['average_cost_for_two'].mean()

if poblaAvgCost > salcedoAvgCost:
    print("Poblacion has more expensive food\n")
elif poblaAvgCost < salcedoAvgCost:
    print("Salcedo Village has more expensive food\n")
else:
    print("Poblacion and Salcedo Village\n")


# Checker
print("Average cost of restos with locality containing \"Poblacion\": " + str(poblaAvgCost))
print("Average cost of restos with locality containing \"Salcedo Village\": " + str(salcedoAvgCost))


Poblacion has more expensive food

Average cost of restos with locality containing "Poblacion": 863.3333333333334
Average cost of restos with locality containing "Salcedo Village": 803.8686131386861


Restaurants in <b><u> Salcedo Village </b></u> has more expensive food